Dev 545 fauna support for by

firehudson opened the pull request.

mmailaender commented on src/lib/stores/document.svelte.ts:
```diff
@@ -281,34 +293,71 @@ export const createDocumentStore = <K extends keyof TypeMapping>(
-					return () => {
-						console.log('delete target', target.id);
-						deleteObject(target.id);
+	const createDocumentHandler = (doc: Record<string, any> = {}) => {
```
Replace with `const createDocumentHandler = (id: string)`

mmailaender replied:
push

firehudson replied:
We can't find document inside createDocumentHandler function from `current` state using below code. Because this is causing call stack error. We need to pass complete document object from parameter instead of document ID

```const existingDoc = $state(current.find((doc) => doc.id === args[0])) || {}); db.byId(args[0]);```

mmailaender commented on src/lib/stores/document.svelte.ts:
```diff
@@ -160,7 +169,11 @@ export const createDocumentStore = <K extends keyof TypeMapping>(
 					return (...args: string[]) => {
-						return getObjects((doc) => doc.id === args[0]).at(0);
+						db.byId(args[0]);
+						const latestData = current.find((doc) => doc.id === args[0]) || {
```
Move into createDocumentHandler

const existingDoc = $state(current.find((doc) => doc.id === args[0])) || {});

firehudson replied:
Can't move `const existingDoc = $state(current.find((doc) => doc.id === args[0])) || {});` into createDocumentHandler. The reason is same as above

mmailaender commented on src/lib/stores/document.svelte.ts:
```diff
@@ -281,34 +293,71 @@ export const createDocumentStore = <K extends keyof TypeMapping>(
+	const createDocumentHandler = (doc: Record<string, any> = {}) => {
+		return {
+			get(target: any, prop: any, receiver: any): any {
+				// Special handling for accessing the whole object
```
Here you need to put the code from byId

```
const existingDoc = $state(current.find((doc) => doc.id === args[0])) || {});
db.byId(args[0]);
```

firehudson replied:
Can't do that, Same reason as above

mmailaender commented on src/lib/stores/document.svelte.ts:
```diff
@@ -281,34 +293,71 @@ export const createDocumentStore = <K extends keyof TypeMapping>(
+				}
+
+				// For regular property access
+				if (prop in doc) {
```
doc becomes `existingDoc`

firehudson commented on src/lib/database/fauna.ts:
```diff
@@ -158,6 +160,26 @@ export const createDatabaseApi = <
+				upsertObjectFromFauna(response.data);
+			}
+		} catch (error) {
+			console.error('Error fetching document from database using byId:', error);
```
We have handled this case in try block. If document not found, fauna returns `cause: "not found"` in response.

Should we need to delete from store if API fails (from catch block)?

![image](https://github.com/user-attachments/assets/b3d793a6-375b-459a-ac9f-194053e1179b)

mmailaender reviewed (changes requested):
Please deal with my comments, resolve them if they are outdated and explain what have change since the last time.

firehudson commented on src/lib/stores/document.svelte.ts:
Yes there was call stack error. Find document by Id inside createDocumentHandler function creates this error. I've explained this in detail in discord channel.

https://discord.com/channels/1222455699146018816/1273168077034360853

firehudson commented:
> Please deal with my comments, resolve them if they are outdated and explain what have change since the last time.

We can't use `current` state inside createDocumentHandler function to find document. This is causing re-rending and call stack error. So I direct pass document from parameter instead of ID.

mmailaender reviewed (changes requested):
I'm getting still the error "RangeError: Maximum call stack size exceeded"

firehudson commented on src/lib/stores/collection.svelte.ts:
```diff
@@ -34,23 +34,22 @@ type CreateCollectionStore = {
 	destroy: () => void;
 };
 
-const createDocumentHandler = (name: string) => {
```
To fix maxsize call stack error, Pass document object instead of name as a parameter

firehudson commented on src/lib/stores/collection.svelte.ts:
```diff
@@ -34,23 +34,22 @@ type CreateCollectionStore = {
+const createDocumentHandler = (doc: Record<string, any> = {}) => {
 	return {
 		get(target: any, prop: any, receiver: any): any {
-			const latestData = $state(getObjects((doc) => doc.name === name).at(0) || {});
```
To fix maxsize call stack error, Remove state from handler function

mmailaender reviewed (changes requested):
With the latest change, the story is almost returned to the start of the story. The whole idea was to have for .byName and .byId lazy loading and with the latest change the lazy loading is rolled back. 

You can check it by cleaning cache and localstorage and reloading.

firehudson commented on src/lib/stores/document.svelte.ts:
```diff
@@ -129,13 +131,19 @@ export const createDocumentStore = <K extends keyof TypeMapping>(
+	const db = createDatabaseApi(client, COLL_NAME, upsertObjectFromFauna, deleteObject);
+	const CollectionStore = createCollectionStore(client);
 
-	const definition = Collection.byName(COLL_NAME);
```
Remove definition variable from `createDocumentStore` function's body and use `Collection.byName(COLL_NAME)` directly at the place of definition variable, please check [here](https://github.com/Etesie/fauna-typed/pull/19/files#diff-3bdb63643a18a300365f108844246e13e2f20707372749f3e2aa5d51d04bcafcR230) for more clarity

firehudson commented on src/lib/stores/document.svelte.ts:
```diff
@@ -229,15 +227,15 @@ export const createDocumentStore = <K extends keyof TypeMapping>(
 					// TODO: Get definition from Fauna
 					// return new Proxy(s.Collection.byName(COLL_NAME), collectionHandler);
-					return definition;
+					return CollectionStore.byName(COLL_NAME);
```
Currently It's working, Could you please check it by taking pull on local and clear local storage?

Although, I'm trying to find alternative of this

mmailaender replied:
If it currently works, then only because you have a lucky race condition. But it can't solve layze fetching conceptual side @firehudson
